akka-stream/src/main/scala/akka/stream/Attributes.scala

```diff
@@ -63,37 +64,36 @@ final case class Attributes private (attributeList: immutable.Seq[Attributes.Att
   /**
    * INTERNAL API
    */
-  private[akka] def nameLifted: Option[String] =
-    if (attributeList.isEmpty)
-      None
-    else {
-      val sb = new java.lang.StringBuilder
-      val iter = attributeList.iterator
-      while (iter.hasNext) {
-        iter.next() match {
-          case Name(name) ⇒
-            if (sb.length == 0) sb.append(name)
-            else sb.append("-").append(name)
-          case _ ⇒
-        }
-      }
-      if (sb.length == 0) None
-      else Some(sb.toString)
-    }
+  private[akka] def nameLifted: Option[String] = Option(nameOrDefault(null))
 
   /**
    * INTERNAL API
    */
-  private[akka] def nameOrDefault(default: String = "unknown-operation"): String = nameLifted match {
-    case Some(name) ⇒ name
-    case _          ⇒ default
-  }
+  private[akka] def nameOrDefault(default: String = "unknown-operation"): String = {
+    @tailrec def concatNames(i: Iterator[Attribute], first: String, buf: StringBuilder): String =
+      if (i.hasNext)
+        i.next() match {
+          case Name(n) ⇒
+            if (buf ne null) concatNames(i, null, buf.append('-').append(n))
+            else if (first ne null) {
+              val b = new StringBuilder(
+                (first.length() + n.length()) match {
```

Comment:
The main improvement is not having to allocate any new builder nor string for when there's only 1 name.
And for the case where there are only 2 names, there will only be 1 allocation of the builders' internal buffer.
The length calculation, to be correct, needs to deal with size overflow etc—hence the match-case. :/